### DISCUSSION TITLE
OpenSSL-3.0: Does HMAC_Init_ex support for variable length key or only standard key?

### DISCUSSION THREAD

**Discussion by ravinjhajharia** [openssl/openssl#22862] [Q&A]
It seems like it used to support in earlier version of openssl?

**Comment by mattcaswell** [comment]
`HMAC_Init_ex` supports a variable length key (although this function is deprecated in 3.0).

**Reply by ravinjhajharia** [reply]
@mattcaswell, thank you for your prompt response. Is EVP_MAC_init the recommended function and does it support variable-length keys?

**Reply by mattcaswell** [reply]
Yes. Please see the documentation here:

https://www.openssl.org/docs/man3.2/man3/EVP_MAC_init.html

**Reply by ravinjhajharia** [reply]
Could you help with a list of variable-length symmetric encryption ciphers which can be used with the above API?

**Reply by t8m** [reply]
It is unclear, what you're asking here because HMAC does not use any cipher inside - just a hash function. And any hash function used in HMAC allows for variable length keys when computing the HMAC.

**Reply by t8m** [reply]
These are the parameters settable for when HMAC is used with EVP_MAC:
https://www.openssl.org/docs/man3.2/man7/EVP_MAC-HMAC.html

**Reply by ravinjhajharia** [reply]
@t8m : My question is related to this code snippet. It uses symmetric AES cipher as provided in run-time arguments to calculate the digest which supports only fixed length key, If I would want to support a variable-length key, then what other alternative can I opt for.

please referer https://www.openssl.org/docs/man3.2/man3/EVP_MAC_init.html

```
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#include <unistd.h>

#include <openssl/evp.h>
#include <openssl/err.h>
#include <openssl/params.h>

int main() {
    EVP_MAC *mac = EVP_MAC_fetch(NULL, getenv("MY_MAC"), NULL);
    const char *cipher = getenv("MY_MAC_CIPHER");
    const char *digest = getenv("MY_MAC_DIGEST");
    const char *key = getenv("MY_KEY");
    EVP_MAC_CTX *ctx = NULL;

    unsigned char buf[4096];
    size_t read_l;
    size_t final_l;

    size_t i;

    OSSL_PARAM params[3];
    size_t params_n = 0;

    if (cipher != NULL)
        params[params_n++] =
            OSSL_PARAM_construct_utf8_string("cipher", (char*)cipher, 0);
    if (digest != NULL)
        params[params_n++] =
            OSSL_PARAM_construct_utf8_string("digest", (char*)digest, 0);
    params[params_n] = OSSL_PARAM_construct_end();

    if (mac == NULL
        || key == NULL
        || (ctx = EVP_MAC_CTX_new(mac)) == NULL
        || !EVP_MAC_init(ctx, (const unsigned char *)key, strlen(key),
                         params))
        goto err;

    while ( (read_l = read(STDIN_FILENO, buf, sizeof(buf))) > 0) {
        if (!EVP_MAC_update(ctx, buf, read_l))
            goto err;
    }

    if (!EVP_MAC_final(ctx, buf, &final_l, sizeof(buf)))
        goto err;

    printf("Result: ");
    for (i = 0; i < final_l; i++)
        printf("%02X", buf[i]);
    printf("\n");

    EVP_MAC_CTX_free(ctx);
    EVP_MAC_free(mac);
    exit(0);

 err:
    EVP_MAC_CTX_free(ctx);
    EVP_MAC_free(mac);
    fprintf(stderr, "Something went wrong\n");
    ERR_print_errors_fp(stderr);
    exit (1);
}
```

**Reply by t8m** [reply]
This example code uses some fictional MY_MAC algorithm. However we are talking about HMAC here which does not use ciphers but only digests internally. The key as passed as a parameter of EVP_MAC_init() can have any length with HMAC.
